lib/core/base/virtual-node.js

```diff
@@ -0,0 +1,79 @@
+// class is unused in the file...
+// eslint-disable-next-line no-unused-vars
+class VirtualNode {
+	/**
+	 * Wrap the real node and provide list of the flattened children
+	 *
+	 * @param node {Node} - the node in question
+	 * @param shadowId {String} - the ID of the shadow DOM to which this node belongs
+	 */
+	constructor(node, shadowId) {
+		this.shadowId = shadowId;
+		this.children = [];
+		this.actualNode = node;
+
+		this._isHidden = null; // will be populated by axe.utils.isHidden
+		this._cache = {};
+
+		// abstract Node and Element APIs so we can run axe in DOM-less
+		// environments. these are static properties in the assumption
+		// that axe does not change any of them while it runs.
+		this.elementNodeType = node.nodeType;
+		this.elementNodeName = node.nodeName.toLowerCase();
+		this.elementId = node.id;
+
+		if (axe._cache.get('nodeMap')) {
+			axe._cache.get('nodeMap').set(node, this);
+		}
+	}
+
+	/**
+	 * Determine if the actualNode has the given class name.
+	 * @see https://j11y.io/jquery/#v=2.0.3&fn=jQuery.fn.hasClass
+	 * @param {String} className - The class to check for.
+	 * @return {Boolean|undefined} True if the actualNode has the given class, false if it does not have the given class, and undefined if the actualNode does not have a className property.
+	 */
+	hasClass(className) {
+		if (typeof this.actualNode.className === 'undefined') {
+			return;
+		}
+
+		let selector = ' ' + className + ' ';
+		return (' ' + this.actualNode.className + ' ').indexOf(selector) >= 0;
+	}
+
+	/**
+	 * Get the value of the given attribute name.
+	 * @param {String} attrName - The name of the attribute.
+	 * @returns {String|undefined} The value of the attribute or undefined if the actualNode does not have a getAttribute function
+	 */
+	attr(attrName) {
+		if (typeof this.actualNode.getAttribute === 'undefined') {
```

Comment:
We expect `.getAttribute` to be a function. If it's somehow set to `{ foo: 'bar' }`, L#55 will throw. Why not do:

```js
if (typeof this.actualNode.getAttribute !== 'function') {
  return null
}
```

Reply:
I guess it depends on how much we care / expect `actualNode` to not be a DOM node. I just need to check that it exists before I call it, but like everywhere else in axe we assume it's a true node and the property won't be overridden with nonsense data.

Reply:
I don't see why adding a guard to protect against unknowns is a bad idea, especially since the `typeof` check is already there.